src/arithmetic/comprehension_funcs/comprehension_funcs.c

```diff
@@ -0,0 +1,139 @@
+/*
+ * Copyright 2018-2020 Redis Labs Ltd. and Contributors
+ *
+ * This file is available under the Redis Labs Source Available License Agreement
+ */
+
+#include "comprehension_funcs.h"
+#include "../../RG.h"
+#include "../func_desc.h"
+#include "../arithmetic_expression.h"
+#include "../../value.h"
+#include "../../util/arr.h"
+#include "../../datatypes/array.h"
+#include "../../execution_plan/record.h"
+#include "../../filter_tree/filter_tree.h"
+
+/* Routine for freeing a list comprehension's subtree of arithmetic expressions.
+ * The predicate and eval routines require special handling to be freed properly. */
+void ListComprehension_Free(AR_ExpNode *exp) {
```

Comment:
I think that if an arithmetic function would have a context, similar to an aggregation function, we could pass the arithmetic expression via this context object, avoiding adding it as a child sub-tree of the list-comprehension function node.

I just don't like the mixing of an arithmetic function defining how to free the structure within it is contained (AR_Exp)